<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>How To Simulate Traits</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">How To Simulate Traits</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(smer)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(genio)</span></code></pre></div>
<p>For a simple illustration, we generate a synthetic allele count
matrix with minor allele frequency <code>maf</code> greater than 5%. The
function <code>simulate_traits()</code> requires the input genotype data
to be stored in the PLINK format (<code>.bim</code>, <code>.bed</code>,
and <code>.fam</code> files). Here we use the package <code>genio</code>
to write the allele count matrix to a PLINK file.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>plink_file <span class="ot">&lt;-</span> <span class="fu">tempfile</span>()</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>n_samples <span class="ot">&lt;-</span> <span class="dv">3000</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>n_snps <span class="ot">&lt;-</span> <span class="dv">5000</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>maf <span class="ot">&lt;-</span> <span class="fl">0.05</span> <span class="sc">+</span> <span class="fl">0.45</span> <span class="sc">*</span> <span class="fu">runif</span>(n_snps)</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>random_minor_allele_counts   <span class="ot">&lt;-</span> (<span class="fu">runif</span>(n_samples <span class="sc">*</span> n_snps) <span class="sc">&lt;</span> maf) <span class="sc">+</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>                                (<span class="fu">runif</span>(n_samples <span class="sc">*</span> n_snps) <span class="sc">&lt;</span> maf)</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>allele_count_matrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(random_minor_allele_counts,</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>  <span class="at">nrow =</span> n_samples,</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>  <span class="at">ncol =</span> n_snps,</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>  <span class="at">byrow =</span> <span class="cn">TRUE</span>,</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>)</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co"># Create .fam and .bim data frames</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>fam <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a>  <span class="at">fam =</span> <span class="fu">sprintf</span>(<span class="st">&quot;F%d&quot;</span>, <span class="dv">1</span><span class="sc">:</span>n_samples),       <span class="co"># Family ID</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a>  <span class="at">id =</span> <span class="fu">sprintf</span>(<span class="st">&quot;I%d&quot;</span>, <span class="dv">1</span><span class="sc">:</span>n_samples),       <span class="co"># Individual ID</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a>  <span class="at">pat =</span> <span class="dv">0</span>,                        <span class="co"># Paternal ID</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a>  <span class="at">mat =</span> <span class="dv">0</span>,                        <span class="co"># Maternal ID</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a>  <span class="at">sex =</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, n_samples, <span class="at">replace =</span> <span class="cn">TRUE</span>),</span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a>  <span class="at">pheno =</span> <span class="sc">-</span><span class="dv">9</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a>)</span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a>bim <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a>  <span class="at">chr =</span> <span class="dv">1</span>,               <span class="co"># Chromosome number</span></span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a>  <span class="at">id =</span> <span class="fu">sprintf</span>(<span class="st">&quot;rs%d&quot;</span>, <span class="dv">1</span><span class="sc">:</span>n_snps),   <span class="co"># SNP name</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a>  <span class="at">posg =</span> <span class="dv">0</span>,                         <span class="co"># Genetic distance (cM)</span></span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a>  <span class="at">pos =</span> <span class="dv">1</span><span class="sc">:</span>n_snps,          <span class="co"># Base-pair position</span></span>
<span id="cb2-28"><a href="#cb2-28" tabindex="-1"></a>  <span class="at">ref =</span> <span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;T&quot;</span>), n_snps, <span class="at">replace =</span> T), <span class="co"># Minor allele</span></span>
<span id="cb2-29"><a href="#cb2-29" tabindex="-1"></a>  <span class="at">alt =</span> <span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;T&quot;</span>), n_snps, <span class="at">replace =</span> T)  <span class="co"># Major allele</span></span>
<span id="cb2-30"><a href="#cb2-30" tabindex="-1"></a>)</span>
<span id="cb2-31"><a href="#cb2-31" tabindex="-1"></a></span>
<span id="cb2-32"><a href="#cb2-32" tabindex="-1"></a><span class="co"># Write to .bed, .fam, and .bim files</span></span>
<span id="cb2-33"><a href="#cb2-33" tabindex="-1"></a><span class="fu">write_plink</span>(</span>
<span id="cb2-34"><a href="#cb2-34" tabindex="-1"></a>  <span class="at">file =</span> plink_file,</span>
<span id="cb2-35"><a href="#cb2-35" tabindex="-1"></a>  <span class="at">X =</span> <span class="fu">t</span>(allele_count_matrix),</span>
<span id="cb2-36"><a href="#cb2-36" tabindex="-1"></a>  <span class="at">fam =</span> fam,</span>
<span id="cb2-37"><a href="#cb2-37" tabindex="-1"></a>  <span class="at">bim =</span> bim</span>
<span id="cb2-38"><a href="#cb2-38" tabindex="-1"></a>)</span></code></pre></div>
<p>To simulate traits, we need to specify the genetic architecture.
E.g., the function requires us to specify</p>
<ul>
<li>the narrow sense heritability <span class="math inline">\(h^2\)</span></li>
<li>the indices SNPs contributing to the narrow sense heritability</li>
<li>the heritability due to epistasis</li>
<li>the indices of SNPs contributing to the epistatic trait
variance</li>
<li>the path to the output file.</li>
</ul>
<p>The simulated trait is written to file in the PLINK phenotype
format.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>pheno_file <span class="ot">&lt;-</span> <span class="fu">tempfile</span>()</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>additive_heritability <span class="ot">&lt;-</span> <span class="fl">0.3</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>gxg_heritability <span class="ot">&lt;-</span> <span class="fl">0.25</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>n_snps_gxg_group <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>n_snps_additive <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>additive_snps <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>n_snps, n_snps_additive, <span class="at">replace =</span> F))</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>gxg_group_1 <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">sample</span>(additive_snps, n_snps_gxg_group, <span class="at">replace =</span> F))</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>gxg_group_2 <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">sample</span>(<span class="fu">setdiff</span>(additive_snps, gxg_group_1),</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>                           n_snps_gxg_group, <span class="at">replace =</span> F))</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="fu">simulate_traits</span>(</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>  plink_file,</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>  pheno_file,</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>  additive_heritability,</span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>  gxg_heritability,</span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>  additive_snps,</span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a>  gxg_group_1,</span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a>  gxg_group_2</span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a>)</span></code></pre></div>
<p>The variance components of the traits are constructed to match the
input. See Crawford et al. (2017) and <a href="https://lcrawlab.github.io/mvMAPIT/articles/tutorial-simulations.html">mvMAPIT
Documentation: Simulate Traits</a> for a full description of the
simulation scheme.</p>
<div id="sessioninfo" class="section level2">
<h2>SessionInfo</h2>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt; Running under: macOS Sequoia 15.2</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co">#&gt; locale:</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co">#&gt; [1] C/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="co">#&gt; time zone: America/New_York</span></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a><span class="co">#&gt; tzcode source: internal</span></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a><span class="co">#&gt; [8] base     </span></span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a><span class="co">#&gt; other attached packages:</span></span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a><span class="co">#&gt;  [1] genio_1.1.2          knitr_1.49           tidyr_1.3.1         </span></span>
<span id="cb4-22"><a href="#cb4-22" tabindex="-1"></a><span class="co">#&gt;  [4] GenomicRanges_1.56.2 GenomeInfoDb_1.40.1  IRanges_2.38.1      </span></span>
<span id="cb4-23"><a href="#cb4-23" tabindex="-1"></a><span class="co">#&gt;  [7] S4Vectors_0.42.1     BiocGenerics_0.50.0  ggplot2_3.5.1       </span></span>
<span id="cb4-24"><a href="#cb4-24" tabindex="-1"></a><span class="co">#&gt; [10] dplyr_1.1.4          smer_0.0.1          </span></span>
<span id="cb4-25"><a href="#cb4-25" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb4-26"><a href="#cb4-26" tabindex="-1"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb4-27"><a href="#cb4-27" tabindex="-1"></a><span class="co">#&gt;  [1] sass_0.4.9              generics_0.1.3          digest_0.6.37          </span></span>
<span id="cb4-28"><a href="#cb4-28" tabindex="-1"></a><span class="co">#&gt;  [4] magrittr_2.0.3          evaluate_1.0.3          grid_4.4.2             </span></span>
<span id="cb4-29"><a href="#cb4-29" tabindex="-1"></a><span class="co">#&gt;  [7] iterators_1.0.14        CompQuadForm_1.4.3      mvtnorm_1.3-3          </span></span>
<span id="cb4-30"><a href="#cb4-30" tabindex="-1"></a><span class="co">#&gt; [10] fastmap_1.2.0           foreach_1.5.2           jsonlite_1.8.9         </span></span>
<span id="cb4-31"><a href="#cb4-31" tabindex="-1"></a><span class="co">#&gt; [13] backports_1.5.0         httr_1.4.7              purrr_1.0.2            </span></span>
<span id="cb4-32"><a href="#cb4-32" tabindex="-1"></a><span class="co">#&gt; [16] UCSC.utils_1.0.0        scales_1.3.0            codetools_0.2-20       </span></span>
<span id="cb4-33"><a href="#cb4-33" tabindex="-1"></a><span class="co">#&gt; [19] jquerylib_0.1.4         cli_3.6.3               rlang_1.1.4            </span></span>
<span id="cb4-34"><a href="#cb4-34" tabindex="-1"></a><span class="co">#&gt; [22] XVector_0.44.0          munsell_0.5.1           withr_3.0.2            </span></span>
<span id="cb4-35"><a href="#cb4-35" tabindex="-1"></a><span class="co">#&gt; [25] cachem_1.1.0            yaml_2.3.10             FMStable_0.1-4         </span></span>
<span id="cb4-36"><a href="#cb4-36" tabindex="-1"></a><span class="co">#&gt; [28] tools_4.4.2             parallel_4.4.2          checkmate_2.3.2        </span></span>
<span id="cb4-37"><a href="#cb4-37" tabindex="-1"></a><span class="co">#&gt; [31] colorspace_2.1-1        GenomeInfoDbData_1.2.12 vctrs_0.6.5            </span></span>
<span id="cb4-38"><a href="#cb4-38" tabindex="-1"></a><span class="co">#&gt; [34] R6_2.5.1                lifecycle_1.0.4         zlibbioc_1.50.0        </span></span>
<span id="cb4-39"><a href="#cb4-39" tabindex="-1"></a><span class="co">#&gt; [37] mvMAPIT_2.0.3           pkgconfig_2.0.3         pillar_1.10.1          </span></span>
<span id="cb4-40"><a href="#cb4-40" tabindex="-1"></a><span class="co">#&gt; [40] bslib_0.8.0             gtable_0.3.6            glue_1.8.0             </span></span>
<span id="cb4-41"><a href="#cb4-41" tabindex="-1"></a><span class="co">#&gt; [43] Rcpp_1.0.14             harmonicmeanp_3.0.1     xfun_0.50              </span></span>
<span id="cb4-42"><a href="#cb4-42" tabindex="-1"></a><span class="co">#&gt; [46] tibble_3.2.1            tidyselect_1.2.1        farver_2.1.2           </span></span>
<span id="cb4-43"><a href="#cb4-43" tabindex="-1"></a><span class="co">#&gt; [49] htmltools_0.5.8.1       rmarkdown_2.29          labeling_0.4.3         </span></span>
<span id="cb4-44"><a href="#cb4-44" tabindex="-1"></a><span class="co">#&gt; [52] compiler_4.4.2</span></span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
